File: Js/frecuencias/frecuencias.js
import frecuenciaAbsoluta from "./frecuenciaAbsoluta.js";
import frecuenciaAbsolutaAcumulada from "./frecuenciaAbsolutaAcumulada.js";
import frecuenciasRelativas from "./frecuenciasRelativas.js";

const labels = ['Cordelia', 'Luna.1999U2', 'Ofelia', 'Luna-1986U10', 'Luna-999U1', 'Bianca', 'Rosalinda', 'Desdémona', 'Calíbano', 'Crésida', 'Belinda', 'Julieta'];
const radius = [13, 15, 16, 20, 20, 22, 27, 29, 30, 33, 34, 42];

const data = [...radius];
const dataLen = data.length;

const rango = data[dataLen - 1] - data[0];
const limites = [10, 20, 30, 40, 50];
const grupos = [[10,20], [20,30], [30,40], [40,50]]
const amplitud = limites[0][1] - limites[0][0]

let marcas = [];

for (let i = 0; i < grupos.length; i++) {
    marcas.push((grupos[i][1] + grupos[i][0]) / 2);
}

let fi = frecuenciaAbsoluta(data, limites)
let fri = frecuenciaAbsolutaAcumulada(data, limites)
console.log(`Agrupados por radio:`, grupos)
console.log(`Marca de clase:`, marcas)
console.log(`Frecuencia Absoluta: ${fi}`)
console.log(`Frecuencia Absoluta Acumulada: ${fri}`) 
console.log(`Frecuencia Relativa: ${frecuenciasRelativas(fi, dataLen)}`)
console.log(`Frecuencia Relativa Acumulada: ${frecuenciasRelativas(fri, dataLen)}`)

File: Python/frecuencias.py
data = [13, 15, 16, 20, 20, 22, 27, 29, 30, 33, 34, 42]
dataL = data.__len__()

rango = data[dataL - 1] - data[0]
limites = [10, 20, 30, 40, 50]
grupos = [[10, 20], [20, 30], [30, 40], [40, 50]]
amplitud = grupos[0][1] - grupos[0][0]

marcas = []

r = grupos.__len__()

for i in range(r):
    marcas.append((grupos[i][1] + grupos[i][0]) / 2)

#Calcula la frecuencia absoluta
def frecuenciaAbsoluta(data, lim):
    llim = len(lim)
    frecuenciaAbsoluta = [0] * llim

    for numero in data:
        for i in range(llim):
            if (numero >= lim[i] and numero < lim[i+1]):
                frecuenciaAbsoluta[i] += 1
                break

    f = []

    for i in range(llim - 1):
        f.append(frecuenciaAbsoluta[i])

    return f

#Calcular la frecuencia absoluta acumulada
def frecuenciaAbsolutaAcumulada(data, lim):
    llim = len(lim)
    frecuenciaAbsoluta = [0] * llim

    for numero in data:
        for i in range(llim):
            if (numero >= lim[i] and numero < lim[i+1]):
                frecuenciaAbsoluta[i] += 1
                break

    frecuenciaAbsolutaAcumulada = []
    acumulada = 0

    for frecuencia in frecuenciaAbsoluta:
        acumulada += frecuencia
        frecuenciaAbsolutaAcumulada.append(acumulada)

    f = []
    
    for i in range(llim -1):
        f.append(frecuenciaAbsolutaAcumulada[i])
    
    return f

def frecuenciasRelativas(abs , dataL):
    f= []
    for freq in abs:
        r = freq / dataL
        r = format(round(r, 2))
        f.append(r)
        
    return f

fi = frecuenciaAbsoluta(data, limites)
Fi = frecuenciaAbsolutaAcumulada(data, limites)

print(f"Agrupadas por radio:", grupos)
print(f"Marca de clase", marcas)
print(f"Frecuencia Absoluta: {fi}")
print(f"Frecuencia Absoluta Acumulada: {Fi}")
print(f"Frecuencia Relativa: {frecuenciasRelativas(fi, dataL)}")
print(f"Frecuencia Relativa Acumulada: {frecuenciasRelativas(Fi, dataL)}")